libraries/wallet/wallet.cpp

```diff
@@ -758,6 +758,131 @@ class wallet_api_impl
 
       return true;
    }
+
+   bool multisig_mode(string on_or_off, fc::optional<string> tx_filename)
+   {
+      if (on_or_off == "on")
+      {
+         if (! tx_filename)
+            FC_THROW("Must provide transaction file name");
+         else
+            _tx_filename = *tx_filename;
+
+         _multisig_mode = true;
+      }
+      else
+         _multisig_mode = false;
+
+      return _multisig_mode;
+   }
+
+   optional<signed_transaction> multisig_import_transaction(string tx_filename)
+   {
+      signed_transaction tmp;
+
+      if ( !fc::exists( tx_filename ) )
+      {
+         elog( "Transaction file not exists" );
+         return optional<signed_transaction>();
+      }
+
+      tmp = fc::json::from_file( tx_filename )
+               .as<signed_transaction>( 2 * GRAPHENE_MAX_NESTED_OBJECTS );
+
+      if ( !tmp.ref_block_num || !tmp.ref_block_prefix || tmp.expiration == fc::time_point_sec() )
+      {
+         elog( "Must fill TaPoS and expiration field in online machine" );
+         return optional<signed_transaction>();
+      }
+      else
+      {
+         _multisig_trx = tmp;
+         return _multisig_trx;
+      }
+   }
+
+   signed_transaction multisig_sign_transaction( const vector<string> &wif_keys, bool broadcast )
+   {
+      // use const reference to call the const variant of `sign()`
+      vector<fc::ecc::private_key> private_keys;
+
+      for ( const string &wif_key : wif_keys )
+      {
+         fc::optional<fc::ecc::private_key> optional_private_key = wif_to_key( wif_key );
+
+         if ( !optional_private_key )
+            FC_THROW( "Invalid private key: ${i}", ( "i", wif_key ) );
+
+         private_keys.push_back(*optional_private_key);
+      }
+
+      for ( const fc::ecc::private_key& privkey : private_keys)
+      {
+         signature_type sig = privkey.sign_compact( _multisig_trx.sig_digest( _chain_id ) );
+         const vector<signature_type> &sigs = _multisig_trx.signatures;
+         if ( std::find( sigs.begin(), sigs.end(), sig ) != sigs.end() )
+         {
+            wlog( "signature of ${fn} had already exist, skip", ( "fn", key_to_wif( privkey ) ) );
```

Comment:
How about "signature of ${fn} already exists, skipping."

Reply:
shame on myself :P, fixed~